File: hackAstone_app/hackAstone_app/JsonPayload.swift
import Foundation

enum JsonPayload {
    static func parse<T: Decodable>(_ raw: String, as type: T.Type = T.self) -> T? {
        var trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if let cut = trimmed.range(of: "}{\"output\"") {
            trimmed = String(trimmed[..<cut.upperBound])
        }
        if let v = tryParse(trimmed, as: T.self) { return v }
        if let range = trimmed.range(of: #"```json\s*"#, options: .regularExpression) {
            let after = trimmed[range.upperBound...]
            if let end = after.range(of: "```") {
                let inner = String(after[..<end.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
                if let v = tryParse(inner, as: T.self) { return v }
            }
        }
        if let range = trimmed.range(of: #"```"#, options: .regularExpression) {
            let after = trimmed[range.upperBound...]
            if let end = after.range(of: "```") {
                let inner = String(after[..<end.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
                if let v = tryParse(inner, as: T.self) { return v }
            }
        }
        if let slice = extractFirstJsonObject(trimmed) {
            return tryParse(slice, as: T.self)
        }
        return nil
    }

    /// 取第一个完整 JSON 对象，避免流式尾包 `{"output":...}` 导致 lastIndex 截错
    private static func extractFirstJsonObject(_ raw: String) -> String? {
        guard let start = raw.firstIndex(of: "{") else { return nil }
        var depth = 0
        var inString = false
        var escape = false
        var i = start
        while i < raw.endIndex {
            let c = raw[i]
            if inString {
                if escape { escape = false }
                else if c == "\\" { escape = true }
                else if c == "\"" { inString = false }
            } else if c == "\"" { inString = true }
            else if c == "{" { depth += 1 }
            else if c == "}" {
                depth -= 1
                if depth == 0 {
                    return String(raw[start ... i])
                }
            }
            i = raw.index(after: i)
        }
        return nil
    }

    private static func tryParse<T: Decodable>(_ text: String, as type: T.Type) -> T? {
        guard let data = text.data(using: .utf8) else { return nil }
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            return nil
        }
    }
}

File: hackAstone_web/src/shared/jsonPayload.ts
/** 提取第一个完整 JSON 对象（避免流式尾包 {@code {"output":...}} 导致 lastIndexOf 截错） */
function extractFirstJsonObject(raw: string): string | null {
  const start = raw.indexOf("{");
  if (start < 0) return null;
  let depth = 0;
  let inString = false;
  let escape = false;
  for (let i = start; i < raw.length; i++) {
    const c = raw[i];
    if (inString) {
      if (escape) escape = false;
      else if (c === "\\") escape = true;
      else if (c === '"') inString = false;
      continue;
    }
    if (c === '"') inString = true;
    else if (c === "{") depth++;
    else if (c === "}") {
      depth--;
      if (depth === 0) return raw.slice(start, i + 1);
    }
  }
  return null;
}

/** 解析 Echo 返回的 JSON（可能含 markdown 围栏、流式尾包或前后缀文案） */
export function parseJsonPayload<T>(raw: string): T | null {
  if (!raw) return null;
  let trimmed = raw.trim();
  const metaCut = trimmed.indexOf('}{"output"');
  if (metaCut > 0) trimmed = trimmed.slice(0, metaCut + 1);
  const direct = tryParse<T>(trimmed);
  if (direct) return direct;
  const fenced =
    trimmed.match(/```json\s*([\s\S]*?)\s*```/i) ||
    trimmed.match(/```\s*([\s\S]*?)\s*```/i);
  if (fenced?.[1]) return tryParse<T>(fenced[1].trim());
  const firstObj = extractFirstJsonObject(trimmed);
  if (firstObj) return tryParse<T>(firstObj);
  return null;
}

function tryParse<T>(text: string): T | null {
  try {
    return JSON.parse(text) as T;
  } catch {
    return null;
  }
}
